designs/0005-static-matrices.md

```diff
@@ -0,0 +1,444 @@
+idea was discarded completely. There is a strategy available now to assign to subsets
+of `varmat` variables that is more efficient than this.
+
+The idea is that when part of a `varmat` is assigned to, the values in that part of the
+`varmat` are saved into the memory arena before being overwritten, and a call is pushed
+onto the chaining stack that will restore the values back from the arena and set the
+respective adjoints of the `varmat` entries to zero.
+
+The assumptions that makes this work is after a subset of a `varmat` is overwritten,
+that part of the `varmat` cannot be used in any further calculations, and so the adjoints
+of that part of the matrix will be zero in the reverse pass at the time of assignment.
+
+This means that assignment into a `varmat` variable would make it so that views
+point into that part of the `varmat` variable's memory no longer point to the expected
+variables.
+
+# Stanc3 implementation
+[stanc3-implementation]: #stanc3-implementation
+
+The stanc3 implementation goal is to implement `varmat` types in a way that
+is invisible to the user. This means the compiler decides how Stan vectors, row
+vectors, and matrices (or arrays of these types) are implemented automatically,
+either picking an appropriate `varmat` or a `matvar` type, and these decisions should
+be made in a way to avoid making programs slower than they would have been
+with just `matvar` types. For brevity, any Stan vector, row vector, or matrix
+type will be referred to from hereon as just a matrix type. Consider a linear
+regression model:
+
+```stan
+data {
+  int N;
+  int K;
+  matrix[N, K] X;
+  vector[N] y;
+}
+
+parameters {
+  vector[K] b;
+  real<lower = 0.0> sigma;
+}
+
+model {
+  vector[N] mu = X * b;
+  target += normal_lpdf(y | mu, sigma);
+}
+```
+
+In this program there are four named matrix variables, `X`, `y`, `b`, and `mu`.
+
+It is immediately evident that because there are no autodiff types in the
+`data` block, all the matrix types in there are implemented with doubles in
+Eigen, so there is no `matvar` or `varmat` to worry about. The same thing
+applies for `transformed data` and the `generated quantities` block.
+
+This leaves `b` and `mu`, matrix variables defined in the `parameters` and
+`model` block. Any matrix variable defined in a `transformed parameters` block
+would be decided the same way.
+
+The algorithm for figuring out the underlying matrix types for these variables
+is as follows:
+
+1. Define all user-defined functions only for `matvar` types, and all
+internally defined variables and returned variables are `matvar` types as well.
+
+2. Assume every named and unnamed matrix variable in the `parameters`,
+`transformed parameters` and `model` blocks use `varmat` types
+
+3. If a matrix variable is used in a function that does not support `varmat`
+arguments, it is made a `matvar` variable.
+
+4. If single elements of a matrix variable are indexed, it is made a `matvar`
+variable.
```

Comment:
Based on the example hierarchical model `varmat` vector supports multi-indexing. So is this _just_ scalar elements; selecting a single row or column does not require `matvar`?

Reply:
Scalar elements also work, they are just slow. Read or write access to scalars leads to slowdowns to a factor of 4. See
https://github.com/stan-dev/design-docs/pull/29#issuecomment-746870561

In the grand scheme of things of a typical model its debatable what effect that would have, but we want to be on the safe side.

Reply:
Good point on row access. I would imagine its faster or close, but a performance test is needed definitely.

Reply:
I'll do a benchmark of row/column access. The overhead here comes from adding a callback for every access. Accessing a row of elements is only one callback, so that should be quick still for varmat compared to matvar.

Reply:
Row/column sums are slower for varmat for size 4 vectors but faster for 8 and up. Columns are easier than rows. Lower on this graph is better for varmat.

![row_column_timings](https://user-images.githubusercontent.com/4742424/103307420-f90e5c00-49dd-11eb-9317-67aa6c30db1b.png)

Reply:
[Here](https://gist.github.com/bbbales2/af6eb405402acc7ac8c7d27a5ea2f56e) is the benchmark code.

Reply:
Thanks. So I would say this is not a problem and row/column access does not rule out varmat.